src/librustc_mir/hair/cx/expr.rs

```diff
@@ -1180,22 +1182,26 @@ fn overloaded_place<'a, 'gcx, 'tcx>(
 
 fn capture_freevar<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
                                    closure_expr: &'tcx hir::Expr,
-                                   freevar: &hir::Freevar,
+                                   index: usize,
+                                   upvar_id: ty::UpvarId,
+                                   def_id: DefId,
                                    freevar_ty: Ty<'tcx>)
                                    -> ExprRef<'tcx> {
-    let var_hir_id = freevar.var_id();
-    let upvar_id = ty::UpvarId {
-        var_path: ty::UpvarPath { hir_id: var_hir_id },
-        closure_expr_id: cx.tcx.hir().local_def_id_from_hir_id(closure_expr.hir_id).to_local(),
-    };
     let upvar_capture = cx.tables().upvar_capture(upvar_id);
     let temp_lifetime = cx.region_scope_tree.temporary_scope(closure_expr.hir_id.local_id);
-    let var_ty = cx.tables().node_type(var_hir_id);
+    let var_ty = cx.tables().node_type(upvar_id.var_path.hir_id);
```

Comment:
I wonder if it'd be better to pass the `var_ty` in as a parameter -- but this seems ok too